server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -2874,8 +2874,11 @@ public String toString() {
     @Override
     public SourceLoader.SyntheticVectorsLoader syntheticVectorsLoader() {
         if (isExcludeSourceVectors) {
-            var syntheticField = new IndexedSyntheticFieldLoader(indexCreatedVersion, fieldType().similarity);
-            return new SyntheticVectorsPatchFieldLoader(syntheticField, syntheticField::copyVectorAsList);
+            return new SyntheticVectorsPatchFieldLoader<>(
+                // Recreate the object for each leaf so that different segments can be searched concurrently.
+                () -> new IndexedSyntheticFieldLoader(indexCreatedVersion, fieldType().similarity),
+                IndexedSyntheticFieldLoader::copyVectorAsList
+            );
```

Comment:
So sneaky!

Why on Earth doesn't the regular synthetic source fall into the same failure? It seems like 
```
 protected SyntheticSourceSupport syntheticSourceSupport() {
```

Would do the same thing as `hasValue` and such are also shared between the leaf value loaders.

Reply:
We do the same kind of trick in [`SourceLoader.Synthetic`](https://github.com/elastic/elasticsearch/blob/c05961f9e3008655abc9c42b1a9b227ce837a18f/server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java#L136). It's sneaky indeed and we should avoid sharing things between these objects but that would require a larger refactoring.

Reply:
🤯 Supplier<SyntheticFieldLoader>

Yikes. Super tricky